### ISSUE TITLE
"invalid clientID" exception is thrown when I want to get information about newly connected client

### ISSUE THREAD

**Issue by Genhis** [planetteamspeak/ts3phpframework#12]
Hello,

I'm using PHP to get notifications about new clients who joined my channel. The problem is, when new client joins the server and I want to get more information about them, the exception TeamSpeak3_Adapter_ServerQuery_Exception("invalid clientID", 0x200) is thrown.

```
$ts3 = TeamSpeak3::factory(...);
TeamSpeak3_Helper_Signal::getInstance()->subscribe("notifyClientmoved", "clientMoved");
$ts3->notifyRegister("channel", 50);

function clientMoved(TeamSpeak3_Adapter_ServerQuery_Event $e, TeamSpeak3_Node_Host $host) {
	$client = $ts3->clientGetById($e->clid);
	//other stuff
}
```

For now, I'm using a workaround:
```
try {
	$client = $ts3->clientGetById($e->clid);
}
catch(TeamSpeak3_Adapter_ServerQuery_Exception $ex) {
	if($ex->getCode() != 0x200)
		throw $ex;
	$ts3->clientListReset();
	$client = $ts3->clientGetById($e->clid);
}
```

But is that really the right way to do it? I would suggest implementing this feature into the framework.

Thanks in advance for any ideas.

**Comment by botorabi** [comment]
Hi,
I think you are on the right path with resetting the client list. In my case I do the same whenever a client related event arrives. I do a similar resetting for channels whenever a channel related event arrives. See here for an example:

snip snip ...

        $TYPE_CHANNEL = "channel";
        $TYPE_CLIENT  = "client";
        $type = $event->getType();
        // check if the event was something about channels
        if (strlen($type) >= strlen($TYPE_CHANNEL) &&
            strcmp(substr($type, 0, strlen($TYPE_CHANNEL)), $TYPE_CHANNEL) === 0) {

            //Log::debug(self::$TAG, " updating the channel list");
            $ts3server->channelListReset();
        }
        if (strlen($type) >= strlen($TYPE_CLIENT) &&
            strcmp(substr($type, 0, strlen($TYPE_CLIENT)), $TYPE_CLIENT) === 0) {

            //Log::debug(self::$TAG, " updating the client list");
            $ts3server->clientListReset();
        }

snip snip ...

**Comment by svenpaulsen** [comment]
When dealing with ServerQuery events, I'd recommend using this code:

```
function clientMoved(TeamSpeak3_Adapter_ServerQuery_Event $e, TeamSpeak3_Node_Host $host) {
    $client = new TeamSpeak3_Node_Client($host->serverGetSelected(), array("clid" => $e->clid));
    
    // your code
}
```

This way, your script would not need to request and reset the entire list of clients all the time. Usually, when I'm creating bot-like scripts, I'm requesting the list of clients just one time right after connecting. After that I'm updating it manually listening for all the important events (e.g. cliententerview, clientleftview, clientmoved, clientupdate, ...).

**Comment by Genhis** [comment]
The reason I need `Client` object is that I want to get their database ID and kick them. And if I read your code correct, then an instance of TeamSpeak3_Node_Client I'll create, will not posess such information, will it?

**Comment by ronindesign** [comment]
@Genhis The new `TeamSpeak3_Node_Client` will be valid and kickable, use:

`$client->kick()`

**Comment by botorabi** [comment]
Hi Sven,
this is indeed a better way than recreating the entire list on every related event. This may be something the library could do, though: evaluating the events and adapting the client and channel lists automatically, it is. I remember I ran into the same problem as Genhis until I noticed that the lists can get out of sync during the lifetime of the app.

**Comment by ronindesign** [comment]
I think forcing native objects to do this management is out of scope of the framework, as some applications may not want this additional overhead by default. Indeed it is a _framework_ not an application or console and should not be required to maintain or update it's own internal state automatically, that could end up being very troublesome for debugging in other situations, framework is only meant to provide a snapshot of state at a point in time.

If this added layer of fault-tolerance is required, it should be implemented only in the scope of application. Either people using the framework could handle this, or another package could be created such as `TeamSpeak3_Console` that is made to manage state, update object lists, and wait by default (e.g. `TeamSpeak3_Adapter_ServerQuery->wait()`)

**Comment by svenpaulsen** [comment]
When you're creating a TeamSpeak3_Node_Client object manually the way I explained - using a valid client ID - the framework will automatically issue the **clientinfo** command when you're trying to access a property that doesn't exist. So I don't think this will be an issue.

**Comment by Genhis** [comment]
Thank you for explanation. So I'll do as you suggest. :)

**Comment by svenpaulsen** [comment]
Alrighty! Please let me know if it's working for you. :-)
